dp/drivers/common.c

```diff
@@ -0,0 +1,326 @@
+{
+	uint16_t out = 0;
+
+#define COPY_AND_RESET(dst, src, bit_out, bit_in) \
+	if ((src) & (bit_in)) { \
+		(dst) |= (bit_out); \
+		(src) &= ~(bit_in); \
+	}
+
+	COPY_AND_RESET(out, in, ETH_RSS_NONFRAG_IPV4_TCP, ETH_RSS_IPV4_TCP);
+	COPY_AND_RESET(out, in, ETH_RSS_NONFRAG_IPV4_UDP, ETH_RSS_IPV4_UDP);
+
+#undef COPY_AND_RESET
+
+	assert(in == 0);
+
+	return out;
+}
+
+static void translate_conf(struct rte_eth_conf *out, struct ix_rte_eth_conf *in)
+{
+	char *p;
+
+	memset(out, 0, sizeof(struct rte_eth_conf));
+
+#define COPY_AND_RESET(field) \
+	do { \
+		out->field = in->field; \
+		in->field = 0; \
+	} while (0)
+
+#define COPY_AND_RESET2(field, f) \
+	do { \
+		out->field = f(in->field); \
+		in->field = 0; \
+	} while (0)
+
+	COPY_AND_RESET2(rx_adv_conf.rss_conf.rss_hf, translate_conf_rss_hf);
+	COPY_AND_RESET(rxmode.header_split);
+	COPY_AND_RESET(rxmode.hw_ip_checksum);
+	COPY_AND_RESET(rxmode.hw_strip_crc);
+	COPY_AND_RESET(rxmode.hw_vlan_filter);
+	COPY_AND_RESET(rxmode.jumbo_frame);
+	COPY_AND_RESET2(rxmode.mq_mode, translate_conf_rxmode_mq_mode);
+	COPY_AND_RESET(rxmode.split_hdr_size);
+	COPY_AND_RESET2(txmode.mq_mode, translate_conf_txmode_mq_mode);
+
+#undef COPY_AND_RESET
+#undef COPY_AND_RESET2
+
+	for (p = (char *) in; p < (char *) in + sizeof(*in); p++)
+		assert(!*p);
+}
+
+struct rte_eth_rxconf rx_conf;
+struct rte_eth_txconf tx_conf;
+
+struct drv_init_tble {
+	char *name;
+	int (*init_fn)(struct ix_rte_eth_dev *dev, const char *driver_name);
+};
+
+struct drv_init_tble drv_init_tbl[] = {
+	{ "rte_ixgbe_pmd", ixgbe_init },
+	{ "rte_ixgbevf_pmd", ixgbe_init },
+	{ NULL, NULL }
+};
+
+int driver_init(struct pci_dev *pci_dev, struct ix_rte_eth_dev **ethp)
+{
+	int i;
+	int ret;
+	struct ix_rte_eth_dev *dev;
+	struct rte_eth_dev_info dev_info;
+	struct rte_eth_conf conf;
+	struct rte_pci_driver *driver;
+	struct drv_init_tble *dpdk_drv;
+	uint8_t port;
+
+	port = rte_eth_dev_find_free_port();
+
+	ret = dpdk_devinit(pci_dev, &driver);
+	if (ret < 0)
+		return ret;
+
+	rte_eth_dev_info_get(port, &dev_info);
+
+	rx_conf = dev_info.default_rxconf;
+	tx_conf = dev_info.default_txconf;
+
+	/* We don't need a dev_private, thus a 1-byte allocation. */
+	dev = eth_dev_alloc(1);
+	dev->port = port;
+	spin_lock_init(&dev->lock);
+
+	translate_conf(&conf, &dev->data->dev_conf);
+
+	conf.fdir_conf.mode = RTE_FDIR_MODE_PERFECT;
+	conf.fdir_conf.pballoc = RTE_FDIR_PBALLOC_256K;
+	conf.fdir_conf.status = RTE_FDIR_REPORT_STATUS;
+	conf.fdir_conf.mask.vlan_tci_mask = 0x0;
+	conf.fdir_conf.mask.ipv4_mask.src_ip = 0xFFFFFFFF;
+	conf.fdir_conf.mask.ipv4_mask.dst_ip = 0xFFFFFFFF;
+	conf.fdir_conf.mask.src_port_mask = 0xFFFF;
+	conf.fdir_conf.mask.dst_port_mask = 0xFFFF;
+	conf.fdir_conf.mask.mac_addr_byte_mask = 0;
+	conf.fdir_conf.mask.tunnel_type_mask = 0;
+	conf.fdir_conf.mask.tunnel_id_mask = 0;
+	conf.fdir_conf.drop_queue = 127;
+	conf.fdir_conf.flex_conf.nb_payloads = 0;
+	conf.fdir_conf.flex_conf.nb_flexmasks = 0;
```

Comment:
We need to check if that how much that `conf` is ixgbe-specific or if the parameters set here are fine for i40e too.